client.go

```diff
@@ -69,151 +45,48 @@ func NewClientWithFiles(gw string, certFile string, keyFile string) (Client, err
 		return Client{}, err
 	}
 
-	return newClientWithConn(gw, conn), nil
+	return newClientWithConn(conn)
 }
 
 func (c *Client) Send(n Notification) error {
-	c.notifs <- n
-	return nil
-}
-
-func (c *Client) reportFailedPush(v interface{}, err *Error) {
-	failedNotif, ok := v.(Notification)
-	if !ok || v == nil {
-		return
+	if c.sess.Disconnected() {
+		c.reconnectAndRequeue()
 	}
 
-	select {
-	case c.FailedNotifs <- NotificationResult{Notif: failedNotif, Err: *err}:
-	default:
-	}
+	return c.sess.Send(n)
 }
 
-func (c *Client) requeue(cursor *list.Element) {
-	// If `cursor` is not nil, this means there are notifications that
-	// need to be delivered (or redelivered)
-	for ; cursor != nil; cursor = cursor.Next() {
-		if n, ok := cursor.Value.(Notification); ok {
-			go func() { c.notifs <- n }()
-		}
-	}
-}
-
-func (c *Client) handleError(err *Error, buffer *buffer) *list.Element {
-	cursor := buffer.Back()
-
-	for cursor != nil {
-		// Get notification
-		n, _ := cursor.Value.(Notification)
+func (c *Client) reconnectAndRequeue() {
+	c.sessm.Lock()
+	defer c.sessm.Unlock()
 
-		// If the notification, move cursor after the trouble notification
-		if n.Identifier == err.Identifier {
-			go c.reportFailedPush(cursor.Value, err)
+	// Pull off undelivered notifications
+	notifs := c.sess.RequeueableNotifications()
 
-			next := cursor.Next()
-
-			buffer.Remove(cursor)
-			return next
-		}
-
-		cursor = cursor.Prev()
-	}
-
-	return cursor
-}
+	// Reconnect
+	c.sess = nil
 
-func (c *Client) runLoop() {
-	sent := newBuffer(50)
-	cursor := sent.Front()
+	for c.sess == nil {
+		sess := newSession(c.conn)
 
-	// APNS connection
-	for {
-		err := c.Conn.Connect()
+		err := sess.Connect()
 		if err != nil {
-			// TODO Probably want to exponentially backoff...
+			// TODO retry policy
+			// TODO connect error channel
+			// Keep trying to connect
```

Comment:
There are cases where retrying is not what we want, such as an invalid or expired certificate.

We need the behaviour to depend on the specific error that occurred in sess.Connect().